<a href="https://colab.research.google.com/github/shivasaich16/AIML---ASSIGNMENT/blob/main/AIML_BASICS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
9
# Install the TensorFlow library using pip.
!pip install tensorflow

 CHAPTER 1 : Introduction to TensorFlow. .

In [ ]:
# Importing the TensorFlow library.
import tensorflow as tf
# Printing the installed TensorFlow version.
print(tf.__version__)

In [ ]:
import tensorflow as tf
import numpy as np
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense

# Define a simple sequential model with one dense layer.
model = Sequential([Dense(units = 1, input_shape = [1])])
# Compile the model using SGD optimizer and mean squared error loss.
model.compile(optimizer = 'sgd', loss = 'mean_squared_error') #calculates mean_square_error
# Define input (xs) and output (ys) training data for a linear relationship.
xs = np.array([-1.0,0.0,1.0,2.0,3.0,4.0], dtype = float)
ys = np.array([-3.0,-1.0,1.0,3.0,5.0,7.0], dtype = float)
# Train the model for 500 epochs.
model.fit(xs,ys, epochs = 500)

# Make a prediction for a new input value (10.0).
print(model.predict(np.array([10.0])))

Epoch 1/500


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 280ms/step - loss: 33.5971
Epoch 2/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 55ms/step - loss: 26.7610
Epoch 3/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 45ms/step - loss: 21.3759
Epoch 4/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - loss: 17.1326
Epoch 5/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - loss: 13.7877
Epoch 6/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 41ms/step - loss: 11.1497
Epoch 7/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - loss: 9.0681
Epoch 8/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 41ms/step - loss: 7.4243
Epoch 9/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 41ms/step - loss: 6.1251
Epoch 10/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 44ms/step - loss: 5.0971
Epoch 11/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - loss: 4.2826
Epoch 12/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - loss: 3.6362
Epoch 13/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 44ms/step - loss: 3.1222
Epoch 14/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 43ms/step - loss: 2.7124
Epoch 15/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 45ms/step - loss: 2.3847
Epoch 16/500
1/1 ━━━━━━━

In [ ]:
from tensorflow.keras import Sequential, Input
from tensorflow.keras.layers import Dense

# Define a sequential model with an explicit Input layer and one Dense layer.
model = Sequential([Input(shape=(1,)), Dense(units=1)])

In [ ]:
# Compile the model using SGD optimizer and mean squared error loss.
model.compile(optimizer ='sgd', loss = 'mean_squared_error')

In [ ]:
# Define input (xs) and output (ys) training data.
xs = np.array([-1.0,0.0,1.0,2.0,3.0,4.0])
ys = np.array([-3.0,-1.0,1.0,3.0,5.0,7.0])
# Train the model for 500 epochs.
model.fit(xs,ys, epochs = 500)

Epoch 1/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 46ms/step - loss: 0.2009
Epoch 2/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - loss: 0.1964
Epoch 3/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 44ms/step - loss: 0.1921
Epoch 4/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 44ms/step - loss: 0.1880
Epoch 5/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 45ms/step - loss: 0.1839
Epoch 6/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - loss: 0.1800
Epoch 7/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - loss: 0.1762
Epoch 8/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - loss: 0.1725
Epoch 9/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 41ms/step - loss: 0.1689
Epoch 10/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - loss: 0.1654
Epoch 11/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 45ms/step - loss: 0.1619
Epoch 12/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 44ms/step - loss: 0.1586
Epoch 13/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 43ms/step - loss: 0.1553
Epoch 14/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 44ms/step - loss: 0.1521
Epoch 15/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 60ms/step - loss: 0.1489
Epoch 16/500
1/1 ━━

In [ ]:
# Make a prediction for a new input value (10.0).
print(model.predict(np.array([10.0])))

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 39ms/step
[[18.992659]]


In [ ]:
import tensorflow as tf
import numpy as np
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense
# Define a dense layer.
l0 = Dense(units = 1, input_shape = [1])
# Create a sequential model with the defined layer.
model = Sequential([l0])
# Compile the model.
model.compile(optimizer = 'sgd', loss = 'mean_squared_error')
# Define training data.
xs = np.array([-1.0,0.0,1.0,2.0,3.0,4.0], dtype = float)
ys = np.array([-3.0,-1.0,1.0,3.0,5.0,7.0], dtype = float)
# Train the model.
model.fit(xs,ys, epochs = 500)
# Make a prediction.
print(model.predict(np.array([10.0])))
# Print the learned weights (parameters) of the layer.
print("Here is what I learned: {}".format(l0.get_weights()))

Epoch 1/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 242ms/step - loss: 34.7416
Epoch 2/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 43ms/step - loss: 27.6662
Epoch 3/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - loss: 22.0929
Epoch 4/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - loss: 17.7013
Epoch 5/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 43ms/step - loss: 14.2397
Epoch 6/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - loss: 11.5098
Epoch 7/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 46ms/step - loss: 9.3557
Epoch 8/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - loss: 7.6549
Epoch 9/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 41ms/step - loss: 6.3107
Epoch 10/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 43ms/step - loss: 5.2472
Epoch 11/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 44ms/step - loss: 4.4047
Epoch 12/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 49ms/step - loss: 3.7362
Epoch 13/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 55ms/step - loss: 3.2047
Epoch 14/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 45ms/step - loss: 2.7811
Epoch 15/500
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - loss: 2.4425
Epoch 16/500

CHAPTER 2:- Introduction to Computer Vision. .


TOPIC 1 : Designing The Neural Network


In [ ]:
import tensorflow as tf
# Define a sequential Keras model for image classification.
model = tf.keras.Sequential([
                           tf.keras.Input(shape=(28,28)), # Input layer specifying image dimensions (28x28 pixels).
                           tf.keras.layers.Flatten(),      # Flattens the 2D input into a 1D array.
                           tf.keras.layers.Dense(128,activation=tf.nn.relu), # Dense hidden layer with 128 neurons and ReLU activation.
                           tf.keras.layers.Dense(10,activation=tf.nn.softmax) # Output layer with 10 neurons (for 10 classes) and Softmax activation for probabilities.
])

In [ ]:
import tensorflow as tf
# Load the Fashion MNIST dataset.
data = tf.keras.datasets.fashion_mnist
(training_images, training_labels), (test_images, test_labels) = data.load_data()

# Normalize pixel values to be between 0 and 1.
training_images = training_images / 255.0;
test_images = test_images / 255.0;

# Define the neural network model.
model = tf.keras.Sequential([
                             tf.keras.Input(shape=(28,28)), # Input layer for 28x28 images.
                             tf.keras.layers.Flatten(),      # Flattens the images into a 1D array.
                             tf.keras.layers.Dense(128,activation=tf.nn.relu), # Hidden dense layer with 128 neurons and ReLU activation.
                             tf.keras.layers.Dense(10,activation=tf.nn.softmax)]) # Output dense layer with 10 neurons and Softmax activation.

# Compile the model with Adam optimizer, sparse categorical crossentropy loss, and accuracy metric.
model.compile(optimizer = 'adam', loss = 'sparse_categorical_crossentropy', metrics = ['accuracy']) # used crossentropy for loss calculation  and adam to optimize and metrics to calculate accuracy
# Train the model for 5 epochs.
model.fit(training_images, training_labels, epochs = 5) # testing it 5 times

Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 2ms/step - accuracy: 0.8253 - loss: 0.4966
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8637 - loss: 0.3778
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8768 - loss: 0.3380
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.8836 - loss: 0.3145
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8897 - loss: 0.2964


In [ ]:
# Assign the Fashion MNIST dataset to a variable for easier access.
data = tf.keras.datasets.fashion_mnist

In [ ]:
# Load the training and test images and labels from the dataset.
(training_images, training_labels), (test_images, test_labels) = data.load_data()
# Normalize the training images by dividing by 255.0.
training_images = training_images / 255.0
# Normalize the test images by dividing by 255.0.
test_images = test_images / 255.0


In [ ]:
# Define the neural network model using a sequential structure.
model = tf.keras.models.Sequential([tf.keras.layers.Flatten(input_shape=(28,28)), # Flatten layer to convert 2D images to 1D, specifying input shape.
tf.keras.layers.Dense(128, activation=tf.nn.relu), # Hidden dense layer with 128 neurons and ReLU activation.
tf.keras.layers.Dense(10, activation=tf.nn.softmax) # Output dense layer with 10 neurons and Softmax activation.
])
# Compile the model with Adam optimizer, sparse categorical crossentropy loss, and accuracy metric.
model.compile(optimizer = 'adam', loss = 'sparse_categorical_crossentropy', metrics = ['accuracy'])
# Train the model on the training data for 5 epochs.
model.fit(training_images, training_labels, epochs = 5)
# Evaluate the model's performance on the test data.
model.evaluate(test_images,test_labels)

/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step - accuracy: 0.8256 - loss: 0.4960
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8655 - loss: 0.3747
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8770 - loss: 0.3348
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8865 - loss: 0.3091
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8913 - loss: 0.2937
313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.8718 - loss: 0.3630


[0.36299920082092285, 0.8718000054359436]

In [ ]:
# Make predictions on the test images.
classifications = model.predict(test_images)
# Print the prediction probabilities for the first test image.
print(classifications[0])
# Print the true label for the first test image.
print(test_labels[0])
# The comment 'it happens due to over fitting' refers to potential reasons for discrepancies between predictions and true labels.

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step
[3.8196191e-07 7.3395699e-07 6.0119959e-08 6.1071368e-08 5.3962492e-07
 1.0995097e-03 1.2388833e-07 6.6478462e-03 1.4866188e-07 9.9225068e-01]
9


In [ ]:
# Import the TensorFlow library.
import tensorflow as tf
# Define a custom Keras Callback to stop training early.
class myCallback(tf.keras.callbacks.Callback):
 # This method is called at the end of each epoch.
 def on_epoch_end(self, epoch, logs={}):
  # Check if the 'accuracy' metric has exceeded 0.95.
  if(logs.get('accuracy')>0.95):
   print("\nReached 95% accuracy so cancelling training!")
   # If accuracy is above 95%, set stop_training to True.
   self.model.stop_training = True
# Create an instance of the custom callback.
callbacks = myCallback()
# Load the Fashion MNIST dataset.
mnist = tf.keras.datasets.fashion_mnist
(training_images, training_labels), (test_images, test_labels) = mnist.load_data()
# Normalize the training images.
training_images=training_images/255.0
# Normalize the test images.
test_images=test_images/255.0
# Define the neural network model.
model = tf.keras.models.Sequential([
 tf.keras.layers.Flatten(), # Flattens input images.
 tf.keras.layers.Dense(128, activation=tf.nn.relu), # Hidden dense layer with ReLU activation.
 tf.keras.layers.Dense(10, activation=tf.nn.softmax) # Output dense layer with Softmax activation.
])
# Compile the model.
model.compile(optimizer='adam',
 loss='sparse_categorical_crossentropy',
 metrics=['accuracy'])
# Train the model for up to 50 epochs, using the custom callback.
model.fit(training_images, training_labels, epochs=50,
 callbacks=[callbacks])

Epoch 1/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 9s 3ms/step - accuracy: 0.8263 - loss: 0.4975
Epoch 2/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step - accuracy: 0.8656 - loss: 0.3763
Epoch 3/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8768 - loss: 0.3347
Epoch 4/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8851 - loss: 0.3118
Epoch 5/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8930 - loss: 0.2921
Epoch 6/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8957 - loss: 0.2777
Epoch 7/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9014 - loss: 0.2659
Epoch 8/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9040 - loss: 0.2565
Epoch 9/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9078 - loss: 0.2465
Epoch 10/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9114 - loss: 0.2388
Epoch 11/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9132 - loss: 0.2313
Epoch 12/50
1875/1875 ━━━━━━━━

CHAPTER 3 DETECTING FEATURES IN IMAGES


1.CONVULATIONS  : A convolution is simply a filter of weights that are used to multiply a pixel with its
neighbors to get a new value for the pixel. For example, consider the ankle boot
image from Fashion MNIST and the pixel values for it as shown



In [ ]:
new_val = (-1*0) + (0*64) + (-2*228) + (.5*48) + (4.5*142) + (2 * 226) + (-3*168)
# This line calculates a new pixel value based on a weighted sum of surrounding pixel values.
# This is a common operation in image processing, particularly in convolutional filters.
# Each term represents a pixel value multiplied by its corresponding weight in the filter.
# -1*0: Weight -1 applied to a pixel with value 0
# 0*64: Weight 0 applied to a pixel with value 64
# -2*228: Weight -2 applied to a pixel with value 228
# .5*48: Weight 0.5 applied to a pixel with value 48
# 4.5*142: Weight 4.5 applied to a pixel with value 142
# 2*226: Weight 2 applied to a pixel with value 226
# -3*168: Weight -3 applied to a pixel with value 168

This equals 577, which will be the new value for that pixel. Repeating this process
across every pixel in the image will give us a filtered image.

Implementing Convulation Neural Network(CNN)

In [ ]:
import tensorflow as tf # Import the TensorFlow library.
data = tf.keras.datasets.fashion_mnist # Access the Fashion MNIST dataset from Keras datasets.
(training_images, training_labels), (test_images, test_labels) = data.load_data() # Load the training and test image data and labels.

training_images = training_images/255.0 # Normalize training images to a 0-1 scale.
test_images = test_images/255.0 # Normalize test images to a 0-1 scale.

# Define a Sequential Keras model.
model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(28,28)), # Flatten the 28x28 input images into a 1D array.
    tf.keras.layers.Dense(128,activation = tf.nn.relu), # Add a dense (fully connected) hidden layer with 128 neurons and ReLU activation.
    tf.keras.layers.Dense(10,activation = tf.nn.softmax) # Add a dense output layer with 10 neurons (for 10 classes) and Softmax activation.
])

# Compile the model with the Adam optimizer, sparse categorical crossentropy loss, and accuracy metric.
model.compile(optimizer = 'adam', loss = 'sparse_categorical_crossentropy', metrics = ['accuracy'])

# Train the model on the training data for 5 epochs.
model.fit(training_images, training_labels, epochs = 5)

29515/29515 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
26421880/26421880 ━━━━━━━━━━━━━━━━━━━━ 1s 0us/step
5148/5148 ━━━━━━━━━━━━━━━━━━━━ 0s 1us/step
4422102/4422102 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 2ms/step - accuracy: 0.8231 - loss: 0.5002
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8644 - loss: 0.3767
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8777 - loss: 0.3366
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8873 - loss: 0.3118
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8922 - loss: 0.2948


using convulaiton layers used as the input layer to neural network


In [ ]:
import tensorflow as tf

tf.keras.layers.Conv2D(64, (3, 3), activation='relu', input_shape=(28, 28, 1))
#Why 64?:

#Feature Extraction: Having 64 filters means that this layer will learn to detect 64 different types of features from the input image.
#Common Practice: Numbers like 32, 64, 128, 256, etc., are powers of 2 and are commonly used as the number of filters in convolutional layers.
#Increasing Complexity: As you go deeper into a convolutional neural network, it's common to increase the number of filters while decreasing the spatial dimensions (e.g., through pooling).
#In essence, choosing 64 filters allows the layer to extract a good number of diverse features from your 28x28x1 input image, which is a common and effective starting point for many image classification tasks.

/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


<Conv2D name=conv2d, built=False>

Pooling`


In [ ]:
tf.keras.layers.MaxPooling2D(2,2)  #pooling layer neural  network

<MaxPooling2D name=max_pooling2d, built=True>

In [ ]:
#using Fasion MNIST(Modified National Institut of standred technology) With CNN
import tensorflow as tf
data = tf.keras.datasets.fashion_mnist
(training_images, training_labels), (test_images, test_labels) = data.load_data()
training_images = training_images.reshape(60000,28,28,1)
training_images = training_images/255.0
test_images = test_images.reshape(10000,28,28,1)
test_images = test_images/255.0
#There are a few things to note here. Remember earlier when I said that the input
#shape for the images had to match what a Conv2D layer would expect, and we updated
#it to be a 28 × 28 × 1 image? The data also had to be reshaped accordingly. 28 × 28 is
#the number of pixels in the image, and 1 is the number of color channels. You’ll typi—
#cally find that this is 1 for a grayscale image or 3 for a color image, where there are
#three channels (red, green, and blue), with the number indicating the intensity of that
#color.
#So, prior to normalizing the images, we also reshape each array to have that extra
#dimension. The following code changes our training dataset from 60,000 images,
#each 28 × 28 (and thus a 60,000 × 28 × 28 array), to 60,000 images, each 28 × 28 × 1:
model = tf.keras.models.Sequential([
                                 tf.keras.layers.Conv2D(64,(3,3),activation = 'relu', input_shape = (28,28,1)),
                                 tf.keras.layers.MaxPooling2D(2,2),
                                 tf.keras.layers.Conv2D(64,(3,3),activation = 'relu'),
                                 tf.keras.layers.MaxPooling2D(2,2),
                                 tf.keras.layers.Flatten(),
                                 tf.keras.layers.Dense(128,activation = tf.nn.relu),
                                 tf.keras.layers.Dense(10,activation = tf.nn.softmax)])
model.compile(optimizer = 'adam', loss = 'sparse_categorical_crossentropy', metrics = ['accuracy'])
model.fit(training_images, training_labels, epochs = 50)

model.evaluate(test_images,test_labels)

classifications = model.predict(test_images)
print(classifications[0])
print(test_labels[0])

#There are a few things to note here. Remember earlier when I said that the input
#shape for the images had to match what a Conv2D layer would expect, and we updated
#it to be a 28 × 28 × 1 image? The data also had to be reshaped accordingly. 28 × 28 is
#the number of pixels in the image, and 1 is the number of color channels. You’ll typi‐
#cally find that this is 1 for a grayscale image or 3 for a color image, where there are
#three channels (red, green, and blue), with the number indicating the intensity of thatcolor.So, prior to normalizing the images, we also reshape each array to have that extradimension. The following code changes our training dataset from 60,000 images,each 28 × 28 (and thus a 60,000 × 28 × 28 array), to 60,000 images, each 28 ×
#28x1

raining_images = training_images.reshape(60000, 28, 28, 1)





/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.8383 - loss: 0.4389
Epoch 2/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.8903 - loss: 0.2966
Epoch 3/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9071 - loss: 0.2498
Epoch 4/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9192 - loss: 0.2174
Epoch 5/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9280 - loss: 0.1906
Epoch 6/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9363 - loss: 0.1674
Epoch 7/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9441 - loss: 0.1476
Epoch 8/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9509 - loss: 0.1294
Epoch 9/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9567 - loss: 0.1130
Epoch 10/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9629 - loss: 0.0984
Epoch 11/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9666 - loss: 0.0875
Epoch 12/50
1875/1875 ━━━━━━━

In [ ]:
import urllib.request
import zipfile
import os

url = "https://storage.googleapis.com/learning-datasets/horse-or-human.zip"
file_name = "horse_or_human.zip"
training_dir = 'horse_or_human_training'

# Create the training directory if it doesn't exist
if not os.path.exists(training_dir):
    os.makedirs(training_dir)

urllib.request.urlretrieve(url, file_name)
zip_ref = zipfile.ZipFile(file_name, 'r')
zip_ref.extractall(training_dir)
zip_ref.close()

from tensorflow.keras.preprocessing.image import ImageDataGenerator
# All images will be rescaled by 1./255 and augmented for better generalization
train_datagen = ImageDataGenerator(
    rescale=1/255,
    rotation_range=40,
    width_shift_range=0.2,
    height_shift_range=0.2,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,
    fill_mode='nearest'
)
train_generator = train_datagen.flow_from_directory(
 training_dir,
 target_size=(300, 300),
 class_mode='binary'
)

Found 1027 images belonging to 2 classes.


horse-or-human/training (we’ll deal with downloading the validation data shortly).
This is the parent directory that will contain subdirectories for the image types.
To use the ImageDataGenerator we now simply use the following code:

In [ ]:
import urllib.request
import zipfile
import os
from tensorflow.keras.preprocessing.image import ImageDataGenerator

#All images will be rescaled by 1./255 and augmented for better generalization
url = "https://storage.googleapis.com/learning-datasets/horse-or-human.zip"
file_name = "horse_or_human.zip"
training_dir = 'horse_or_human_training'

# Create the training directory if it doesn't exist
if not os.path.exists(training_dir):
    os.makedirs(training_dir)

urllib.request.urlretrieve(url, file_name)
zip_ref = zipfile.ZipFile(file_name, 'r')
zip_ref.extractall(training_dir)
zip_ref.close()

train_datagen = ImageDataGenerator(
    rescale=1/255,
    rotation_range=40,
    width_shift_range=0.2,
    height_shift_range=0.2,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,
    fill_mode='nearest'
)
train_generator = train_datagen.flow_from_directory(
training_dir,
target_size = (300,300),
class_mode = 'binary'
)

Found 1027 images belonging to 2 classes.


CNN Architecture for Horses or Humans

In [ ]:
import tensorflow as tf

model = tf.keras.models.Sequential([
    # Input layer explicitly defining the shape of the input images (300x300 pixels with 3 color channels).
    tf.keras.layers.Input(shape=(300, 300, 3)),
    # First convolutional layer with 16 filters, a 3x3 kernel, and ReLU activation.
    tf.keras.layers.Conv2D(16, (3,3), activation='relu'),
    # First max pooling layer, reducing spatial dimensions by 2x2.
    tf.keras.layers.MaxPooling2D(2, 2),
    # Second convolutional layer with 32 filters, a 3x3 kernel, and ReLU activation.
    tf.keras.layers.Conv2D(32, (3,3), activation='relu'),
    # Second max pooling layer.
    tf.keras.layers.MaxPooling2D(2,2),
    # Third convolutional layer with 64 filters, a 3x3 kernel, and ReLU activation.
    tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
    # Third max pooling layer.
    tf.keras.layers.MaxPooling2D(2,2),
    # Fourth convolutional layer with 64 filters, a 3x3 kernel, and ReLU activation.
    tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
    # Fourth max pooling layer.
    tf.keras.layers.MaxPooling2D(2,2),
    # Fifth convolutional layer with 64 filters, a 3x3 kernel, and ReLU activation.
    tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
    # Fifth max pooling layer.
    tf.keras.layers.MaxPooling2D(2,2),
    # Flattens the 3D output from the convolutional layers into a 1D vector for the dense layers.
    tf.keras.layers.Flatten(),
    # Dense (fully connected) hidden layer with 512 neurons and ReLU activation.
    tf.keras.layers.Dense(512, activation='relu'),
    # Output dense layer with 1 neuron and sigmoid activation for binary classification (horse or human).
    tf.keras.layers.Dense(1, activation='sigmoid')
])

The model has been defined. The next step is to compile it.

In [ ]:
# Compile the model with the Adam optimizer, binary crossentropy loss, and accuracy as the evaluation metric.
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
# Display a summary of the model's architecture, including the number of layers, output shapes, and trainable parameters.
model.summary()

Model: "sequential_4"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_20 (Conv2D)              │ (None, 298, 298, 16)   │           448 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_20 (MaxPooling2D) │ (None, 149, 149, 16)   │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_21 (Conv2D)              │ (None, 147, 147, 32)   │         4,640 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_21 (MaxPooling2D) │ (None, 73, 73, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_22 (Conv2D)              │ (None, 71, 71, 64)     │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_22 (MaxPooling2D) │ (None, 35, 35, 64)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_23 (Conv2D)              │ (None, 33, 33, 64)     │        36,928 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_23 (MaxPooling2D) │ (None, 16, 16, 64)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_24 (Conv2D)              │ (None, 14, 14, 64)     │        36,928 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_24 (MaxPooling2D) │ (None, 7, 7, 64)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_4 (Flatten)             │ (None, 3136)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_8 (Dense)                 │ (None, 512)            │     1,606,144 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_9 (Dense)                 │ (None, 1)              │           513 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 1,704,097 (6.50 MB)

 Trainable params: 1,704,097 (6.50 MB)

 Non-trainable params: 0 (0.00 B)

Note how, by the time the data has gone through all the convolutional and pooling
layers, it ends up as 7 × 7 items. The theory is that these will be activated feature maps
that are relatively simple, containing just 49 pixels. These feature maps can then be
passed to the dense neural network to match them to the appropriate labels.
This, of course, leads us to have many more parameters than the previous network, so
it will be slower to train. With this architecture, we’re going to learn 1.7 million
parameters To train the network, we’ll have to compile it with a loss function and an optimizer. In
this case the loss function can be binary cross entropy loss function binary cross
entropy, because there are only two classes, and as the name suggests this is a loss
function that is designed for that scenario. And we can try a new optimizer, root mean
square propagation (RMSprop), that takes a learning rate (lr) parameter that allows us
to tweak the learning. Here’s the code:

In [ ]:
model.compile(
    loss='binary_crossentropy', # Specifies the loss function for binary classification problems.
    optimizer=tf.keras.optimizers.RMSprop(learning_rate=0.001), # Sets the RMSprop optimizer with a learning rate of 0.001.
    metrics=['accuracy'] # Defines the metric(s) to evaluate the model's performance during training and testing.
)

We train by using fit_generator and passing it the training_generator we created
earlier:

In [ ]:
history = model.fit(
    train_generator,
    epochs=15,
    validation_data=validation_generator,
    validation_steps=len(validation_generator)
)

Epoch 1/15
33/33 ━━━━━━━━━━━━━━━━━━━━ 30s 922ms/step - accuracy: 0.7575 - loss: 0.7084 - val_accuracy: 0.6406 - val_loss: 1.5179
Epoch 2/15
33/33 ━━━━━━━━━━━━━━━━━━━━ 27s 830ms/step - accuracy: 0.8569 - loss: 0.3454 - val_accuracy: 0.8008 - val_loss: 1.2551
Epoch 3/15
33/33 ━━━━━━━━━━━━━━━━━━━━ 30s 901ms/step - accuracy: 0.8724 - loss: 0.3506 - val_accuracy: 0.7812 - val_loss: 1.4646
Epoch 4/15
33/33 ━━━━━━━━━━━━━━━━━━━━ 27s 843ms/step - accuracy: 0.8997 - loss: 0.2700 - val_accuracy: 0.8008 - val_loss: 0.9514
Epoch 5/15
33/33 ━━━━━━━━━━━━━━━━━━━━ 26s 800ms/step - accuracy: 0.9075 - loss: 0.2867 - val_accuracy: 0.8320 - val_loss: 0.9987
Epoch 6/15
33/33 ━━━━━━━━━━━━━━━━━━━━ 27s 809ms/step - accuracy: 0.9182 - loss: 0.2078 - val_accuracy: 0.7969 - val_loss: 1.2014
Epoch 7/15
33/33 ━━━━━━━━━━━━━━━━━━━━ 27s 805ms/step - accuracy: 0.9484 - loss: 0.1815 - val_accuracy: 0.7891 - val_loss: 1.7016
Epoch 8/15
33/33 ━━━━━━━━━━━━━━━━━━━━ 27s 806ms/step - accuracy: 0.9172 - loss: 0.2345 - val_accu

Testing Horse or Human Images
It’s all very well to be able to build a model, but of course you want to try it out. A
major frustration of mine when I was starting my AI journey was that I could find
lots of code that showed me how to build models, and charts of how those models
were performing, but very rarely was there code to help me kick the tires of the
model myself to try it out. I’ll try to avoid that in this book!
Testing the model is perhaps easiest using Colab. I’ve provided a Horses or Humans
notebook on GitHub that you can open directly in Colab.
Once you’ve trained the model, you’ll see a section called “Running the Model.”
Before running it, find a few pictures of horses or humans online and download them
to your computer. Pixabay.com is a really good site to check out for royalty-free
images. It’s a good idea to get your test images together first, because the node can
time out while you’re searching.
Figure 3-9 shows a few pictures of horses and humans that I downloaded from Pixa‐
bay to test the model.

In [ ]:
#using the numpy module
import numpy as np
from google.colab import files # access the files from our laptop and post it into colab
from keras.preprocessing import image  # used to preprocess the image
import tensorflow as tf # Import TensorFlow for tensor conversion

uploaded = files.upload() # retrives the uploaded image

for fn in uploaded.keys():
  # predicting images
  path = '/content/' + fn
  img = image.load_img(path, target_size=(300,300))
  x = image.img_to_array(img)
  x = np.expand_dims(x, axis = 0)
  images = np.vstack([x])

  # Explicitly convert to TensorFlow tensor
  images_tensor = tf.convert_to_tensor(images)

  classes = model.predict(images_tensor)
  print(classes)
  print(classes[0])
  if classes[0]>0.5:
    print(fn + " is a human")
  else:
    print(fn + " is a horse")

Saving image1 to image1 (5)
Saving image2 to image2 (3)
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 39ms/step
[[0.]]
[0.]
image1 (5) is a horse
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 39ms/step
[[1.]]
[1.]
image2 (3) is a human


In [ ]:
img = image.load_img(path, target_size=(300, 300)) # Loads the image from the specified 'path' and resizes it to 300x300 pixels.
x = image.img_to_array(img) # Converts the loaded image into a NumPy array.
x = np.expand_dims(x, axis=0) # Adds an extra dimension to the array, which is required as the model expects a batch of images.

In [ ]:
image_tensor = np.vstack([x]) # Stacks arrays in sequence vertically; here, it essentially converts 'x' to a tensor if it isn't already, preparing it for model prediction.

In [ ]:
classes = model.predict(image_tensor) # Uses the trained model to predict the class (horse or human) of the input image_tensor.

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 40ms/step


In [ ]:
if classes[0]>0.5:
  print(fn + " is a human") # If the model's prediction (first element of 'classes') is greater than 0.5, it's classified as a human.
else:
  print(fn + " is a horse") # Otherwise, it's classified as a horse.

image2 (3) is a human


There are a few important points to consider here. First, even though the network
was trained on synthetic, computer-generated imagery, it performs quite well at spot‐
ting horses or humans in real photographs. This is a potential boon in that you may
not need thousands of photographs to train a model, and can do it relatively cheaply
with CGI.
But this dataset also demonstrates a fundamental issue you will face. Your training set
cannot hope to represent every possible scenario your model might face in the wild,
and thus the model will always have some level of overspecialization toward the train‐
ing set. A clear and simple example of that was shown here, where the human in the
center of Figure 3-9 was miscategorized. The training set didn’t include a human in that pose, and thus the model didn’t “learn” that a human could look like that. As a
result, there was every chance it might see the figure as a horse, and in this case, it
did.
What’s the solution? The obvious one is to add more training data, with humans in
that particular pose and others that weren’t initially represented. That isn’t always
possible, though. Fortunately, there’s a neat trick in TensorFlow that you can use to
virtually extend your dataset—it’s called image augmentation, and we’ll explore that
next.

Image Augmentation
In the previous section, you built a horse-or-human classifier model that was trained
on a relatively small dataset. As a result, you soon began to hit problems classifying
some previously unseen images, such as the miscategorization of a woman with a
horse because the training set didn’t include any images of people in that pose.
One way to deal with such problems is with image augmentation. The idea behind
this technique is that, as TensorFlow is loading your data, it can create additional new
data by amending what it has using a number of transforms. For example, take a look
at Figure 3-11. While there is nothing in the dataset that looks like the woman on the
right, the image on the left is somewhat similar.


In a similar way, you can broaden the training set with a variety of other transforma‐
tions, including:
• Rotation
• Shifting horizontally
• Shifting vertically
• Shearing
• Zooming
• Flipping
Because you’ve been using the ImageDataGenerator to load the images, you’ve seen it
do a transform already—when it normalized the images like this:


In [ ]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator
train_datagen = ImageDataGenerator(rescale=1/255) # Initializes an ImageDataGenerator to rescale pixel values to a [0, 1] range.

The other transforms are easily available within the ImageDataGenerator too, so, for
example, you could do something like this

In [ ]:
train_datagen = ImageDataGenerator(
rescale=1./255, # Rescales pixel values from [0, 255] to [0, 1] for normalization.
rotation_range=40, # Randomly rotates images by up to 40 degrees.
width_shift_range=0.2, # Randomly shifts images horizontally by up to 20% of the total width.
height_shift_range=0.2, # Randomly shifts images vertically by up to 20% of the total height.
shear_range=0.2, # Applies random shear transformations.
zoom_range=0.2, # Applies random zoom transformations.
horizontal_flip=True, # Randomly flips images horizontally.
fill_mode='nearest' # Fills in newly created pixels (after rotation or shift) using the nearest pixel values.
)

Here, as well as rescaling the image to normalize it, you’re also doing the following:
• Rotating each image randomly up to 40 degrees left or right
• Translating the image up to 20% vertically or horizontally
• Shearing the image by up to 20%
• Zooming the image by up to 20%
• Randomly flipping the image horizontally or vertically
• Filling in any missing pixels after a move or shear with nearest neighbors


When you retrain with these parameters, one of the first things you’ll notice is that
training takes longer because of all the image processing. Also, your model’s accuracy
may not be as high as it was previously, because previously it was overfitting to a
largely uniform set of data.
In my case, when training with these augmentations my accuracy went down from
99% to 85% after 15 epochs, with validation slightly higher at 89%. (This indicates
that the model is underfitting slightly, so the parameters could be tweaked a bit.)
What about the image from Figure 3-9 that it misclassified earlier? This time, it gets it
right. Thanks to the image augmentations, the training set now has sufficient cover‐
age for the model to understand that this particular image is a human too (see
Figure 3-13). This is just a single data point, and may not be representative of the
results for real data, but it’s a small step in the right direction.

Transfer learning

In [ ]:
from tensorflow.keras.applications.inception_v3 import InceptionV3

# Using weights='imagenet' to automatically download pre-trained weights
pre_trained_model = InceptionV3(input_shape=(150, 150, 3), # Define the input image shape expected by the model (width, height, channels).
                                include_top=False, # Exclude the top (classification) layer of the pre-trained model.
                                weights='imagenet') # Load weights pre-trained on the ImageNet dataset automatically.

# No need to load weights from a file manually as 'imagenet' handles it

In [ ]:
pre_trained_model.summary()

Model: "inception_v3"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer_9       │ (None, 150, 150,  │          0 │ -                 │
│ (InputLayer)        │ 3)                │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2d_401 (Conv2D) │ (None, 74, 74,    │        864 │ input_layer_9[0]… │
│                     │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 74, 74,    │         96 │ conv2d_401[0][0]  │
│ (BatchNormalizatio… │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ activation_376      │ (None, 74, 74,    │          0 │ batch_normalizat… │
│ (Activation)        │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2d_402 (Conv2D) │ (None, 72, 72,    │      9,216 │ activation_376[0… │
│                     │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 72, 72,    │         96 │ conv2d_402[0][0]  │
│ (BatchNormalizatio… │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ activation_377      │ (None, 72, 72,    │          0 │ batch_normalizat… │
│ (Activation)        │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2d_403 (Conv2D) │ (None, 72, 72,    │     18,432 │ activation_377[0… │
│                     │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 72, 72,    │        192 │ conv2d_403[0][0]  │
│ (BatchNormalizatio… │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ activation_378      │ (None, 72, 72,    │          0 │ batch_normalizat… │
│ (Activation)        │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ max_pooling2d_41    │ (None, 35, 35,    │          0 │ activation_378[0… │
│ (MaxPooling2D)      │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2d_404 (Conv2D) │ (None, 35, 35,    │      5,120 │ max_pooling2d_41… │
│                     │ 80)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 35, 35,    │        240 │ conv2d_404[0][0]  │
│ (BatchNormalizatio… │ 80)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ activation_379      │ (None, 35, 35,    │          0 │ batch_normalizat… │
│ (Activation)        │ 80)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2d_405 (Conv2D) │ (None, 33, 33,    │    138,240 │ activation_379[0… │
│                     │ 192)              │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 33, 33,    │        576 │ conv2d_405[0][0]  │
│ (BatchNormalizatio… │ 192)              │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ activation_380      │ (None, 33, 33,    │          0 │ batch_normalizat

 Total params: 21,802,784 (83.17 MB)

 Trainable params: 21,768,352 (83.04 MB)

 Non-trainable params: 34,432 (134.50 KB)

Be warned—it’s huge! Still, take a look through it to see the layers and their names. I
like to use the one called mixed7 because its output is nice and small—7 × 7 images—
but feel free to experiment with others.
Next, we’ll freeze the entire network from retraining and then set a variable to point
at mixed7’s output as where we want to crop the network up to. We can do that with
this code:

In [ ]:
from tensorflow.keras.applications.inception_v3 import InceptionV3

# Using weights='imagenet' to automatically download pre-trained weights
pre_trained_model = InceptionV3(input_shape=(150, 150, 3), # Define the input image shape expected by the model (width, height, channels).
                                include_top=False, # Exclude the top (classification) layer of the pre-trained model.
                                weights='imagenet') # Load weights pre-trained on the ImageNet dataset automatically.

# No need to load weights from a file manually as 'imagenet' handles it

# Freeze all the layers in the pre-trained model
# This prevents the existing weights from being updated during training
for layer in pre_trained_model.layers:
    layer.trainable = False

# Get the layer named 'mixed7' from the model
# Usually used for transfer learning in models like InceptionV3
last_layer = pre_trained_model.get_layer('mixed7')

# Print the output shape of the selected layer
# Helps to understand the feature map dimensions
print("Last layer output shape:", last_layer.output.shape)

# Store the output of the selected layer
# This output will be used to add custom layers on top
last_output = last_layer.output

87910968/87910968 ━━━━━━━━━━━━━━━━━━━━ 1s 0us/step
Last layer output shape: (None, 7, 7, 768)


Note that we print the output shape of the last layer, and you’ll see that we’re getting
7 × 7 images at this point. This indicates that by the time the images have been fed
through to mixed7, the output images from the filters are 7 × 7 in size, so they’re
pretty easy to manage. Again, you don’t have to choose that specific layer; you’re wel‐
come to experiment with others.
Let’s now see how to add our dense layers underneath this:

In [ ]:
# Flatten the output layer to 1 dimension
x = tf.keras.layers.Flatten()(last_output)
# Add a fully connected layer with 1,024 hidden units and ReLU activation
x = tf.keras.layers.Dense(1024, activation='relu')(x)
# Add a final sigmoid layer for classification
x = tf.keras.layers.Dense(1, activation='sigmoid')(x)

It’s as simple as creating a flattened set of layers from the last output, because we’ll be
feeding the results into a dense layer. We then add a dense layer of 1,024 neurons, and
a dense layer with 1 neuron for our output.
Now we can define our model simply by saying it’s our pretrained model’s input fol‐
lowed by the x we just defined. We then compile it in the usual way:

In [ ]:
# Create the new model by connecting the input of the pre-trained model
# to the output of our custom classification layers.
model = tf.keras.Model(pre_trained_model.input, x)

# Compile the new model with the RMSprop optimizer, binary crossentropy loss,
# and accuracy as the evaluation metric. A learning rate of 0.0001 is used.
model.compile(
    optimizer=tf.keras.optimizers.RMSprop(learning_rate=0.0001),
    loss='binary_crossentropy',
    metrics=['accuracy']
)

Multiclass Classification
In all of the examples so far you’ve been building binary classifiers—ones that choose
between two options (horses or humans, cats or dogs). When building multiclass
classifiers the models are almost the same, but there are a few important differences.
Instead of a single neuron that is sigmoid-activated, or two neurons that are binary-
activated, your output layer will now require n neurons, where n is the number of
classes you want to classify. You’ll also have to change your loss function to an appro‐
priate one for multiple categories. For example, whereas for the binary classifiers
you’ve built so far in this chapter your loss function was binary cross entropy, if you
want to extend the model for multiple classes you should instead use categorical cross
entropy. If you’re using the ImageDataGenerator to provide your images the labeling
is done automatically, so multiple categories will work the same as binary ones—the
ImageDataGenerator will simply label based on the number of subdirectories.
Consider, for example, the game Rock Paper Scissors. If you wanted to train a dataset
to recognize the different hand gestures, you’d need to handle three categories. Fortu‐
nately, there’s a simple dataset you can use for this.
There are two downloads: a training set of many diverse hands, with different sizes,
shapes, colors, and details such as nail polish; and a testing set of equally diverse
hands, none of which are in the training set.
You can see some examples in Figure 3-19.

In [ ]:
import zipfile
from tensorflow.keras.preprocessing.image import ImageDataGenerator

# Download the Rock Paper Scissors dataset from a Google Cloud Storage URL.
# The '--no-check-certificate' flag is used to bypass SSL certificate verification.
# The downloaded file will be saved as '/tmp/rps.zip'.
!wget --no-check-certificate \
https://storage.googleapis.com/download.tensorflow.org/data/rps.zip \
-O /tmp/rps.zip

# Define the local path to the downloaded zip file.
local_zip = '/tmp/rps.zip'
# Open the zip file in read mode.
zip_ref = zipfile.ZipFile(local_zip, 'r')
# Extract all the contents of the zip file to the '/tmp/' directory.
zip_ref.extractall('/tmp/')
# Close the zip file.
zip_ref.close()

# Define the directory where the training images are located after extraction.
TRAINING_DIR = "/tmp/rps/"
# Initialize an ImageDataGenerator for data augmentation and preprocessing.
training_datagen = ImageDataGenerator(
rescale = 1./255, # Rescale pixel values from [0, 255] to [0, 1].
rotation_range=40, # Randomly rotate images by up to 40 degrees.
width_shift_range=0.2, # Randomly shift images horizontally by up to 20% of the width.
height_shift_range=0.2, # Randomly shift images vertically by up to 20% of the height.
shear_range=0.2, # Apply random shear transformations.
zoom_range=0.2, # Apply random zoom transformations.
horizontal_flip=True, # Randomly flip images horizontally.
fill_mode='nearest' # Fill any newly created pixels with the nearest available pixel values.
)

--2026-05-25 05:58:53--  https://storage.googleapis.com/download.tensorflow.org/data/rps.zip
Resolving storage.googleapis.com (storage.googleapis.com)... 142.251.2.207, 74.125.137.207, 142.250.101.207, ...
Connecting to storage.googleapis.com (storage.googleapis.com)|142.251.2.207|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 200682221 (191M) [application/zip]
Saving to: ‘/tmp/rps.zip’

/tmp/rps.zip        100%[===================>] 191.38M   119MB/s    in 1.6s    

2026-05-25 05:58:55 (119 MB/s) - ‘/tmp/rps.zip’ saved [200682221/200682221]



Note, however, that when you set up the data generator from this, you have to specify
that the class mode is categorical in order for the ImageDataGenerator to use more
than two subdirectories:

In [ ]:
train_generator = training_datagen.flow_from_directory(
TRAINING_DIR,
target_size=(150,150),
class_mode='categorical'
)

Found 2520 images belonging to 3 classes.


When defining your model, while keeping an eye on the input and output layers, you
want to ensure that the input matches the shape of the data (in this case 150 × 150)
and that the output matches the number of classes (now three):

In [ ]:
model = tf.keras.models.Sequential([
# Note the input shape is the desired size of the image:
# 150x150 with 3 bytes color
# This is the first convolution
tf.keras.layers.Conv2D(64, (3,3), activation='relu',
input_shape=(150, 150, 3)),
tf.keras.layers.MaxPooling2D(2, 2),
# The second convolution
tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
tf.keras.layers.MaxPooling2D(2,2),
# The third convolution
tf.keras.layers.Conv2D(128, (3,3), activation='relu'),
tf.keras.layers.MaxPooling2D(2,2),
# The fourth convolution
tf.keras.layers.Conv2D(128, (3,3), activation='relu'),
tf.keras.layers.MaxPooling2D(2,2),
# Flatten the results to feed into a DNN
tf.keras.layers.Flatten(),
# 512 neuron hidden layer
tf.keras.layers.Dense(512, activation='relu'),
tf.keras.layers.Dense(3, activation='softmax')
])

/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Finally, when compiling your model, you want to ensure that it uses a categorical loss
function, such as categorical cross entropy. Binary cross entropy will not work with
more than two classes:

In [ ]:
model.compile(loss = 'categorical_crossentropy', optimizer='rmsprop',
metrics=['accuracy'])

Training is then the same as before:

In [ ]:
history = model.fit(train_generator, epochs=25,
validation_data = validation_generator, verbose = 1)

Epoch 1/25
79/79 ━━━━━━━━━━━━━━━━━━━━ 29s 330ms/step - accuracy: 0.8742 - loss: 0.3448 - val_accuracy: 0.8468 - val_loss: 0.2916
Epoch 2/25
79/79 ━━━━━━━━━━━━━━━━━━━━ 24s 300ms/step - accuracy: 0.9155 - loss: 0.2402 - val_accuracy: 0.9892 - val_loss: 0.0591
Epoch 3/25
79/79 ━━━━━━━━━━━━━━━━━━━━ 24s 307ms/step - accuracy: 0.9373 - loss: 0.1788 - val_accuracy: 0.9651 - val_loss: 0.0899
Epoch 4/25
79/79 ━━━━━━━━━━━━━━━━━━━━ 23s 294ms/step - accuracy: 0.9377 - loss: 0.1799 - val_accuracy: 0.9005 - val_loss: 0.2373
Epoch 5/25
79/79 ━━━━━━━━━━━━━━━━━━━━ 24s 307ms/step - accuracy: 0.9631 - loss: 0.1138 - val_accuracy: 0.8091 - val_loss: 0.5325
Epoch 6/25
79/79 ━━━━━━━━━━━━━━━━━━━━ 24s 302ms/step - accuracy: 0.9635 - loss: 0.1080 - val_accuracy: 0.9758 - val_loss: 0.0577
Epoch 7/25
79/79 ━━━━━━━━━━━━━━━━━━━━ 24s 301ms/step - accuracy: 0.9679 - loss: 0.0942 - val_accuracy: 0.9651 - val_loss: 0.1125
Epoch 8/25
79/79 ━━━━━━━━━━━━━━━━━━━━ 24s 305ms/step - accuracy: 0.9698 - loss: 0.0980 - val_accu

Your code for testing predictions will also need to change somewhat. There are now
three output neurons, and they will output a value close to 1 for the predicted class,
and close to 0 for the other classes. Note that the activation function used is softmax,
which will ensure that all three predictions will add up to 1. For example, if the model
sees something it’s really unsure about it might output .4, .4, .2, but if it sees some‐
thing it’s quite sure about you might get .98, .01, .01.
Note also that when using the ImageDataGenerator, the classes are loaded in alpha‐
betical order—so while you might expect the output neurons to be in the order of the
name of the game, the order in fact will be Paper, Rock, Scissors.

In [ ]:
import numpy as np
from google.colab import files
from keras.preprocessing import image
import tensorflow as tf # Import TensorFlow for model definition and layers

# Define the Rock Paper Scissors model architecture
# This model expects 150x150 pixel images with 3 color channels
model = tf.keras.models.Sequential([
    tf.keras.layers.Conv2D(64, (3,3), activation='relu', input_shape=(150, 150, 3)),
    tf.keras.layers.MaxPooling2D(2, 2),
    tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2,2),
    tf.keras.layers.Conv2D(128, (3,3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2,2),
    tf.keras.layers.Conv2D(128, (3,3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2,2),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(512, activation='relu'),
    tf.keras.layers.Dense(3, activation='softmax') # Output layer for 3 classes (Paper, Rock, Scissors)
])

# Compile the model
# Assuming this model was previously trained; if not, predictions will be random.
model.compile(loss='categorical_crossentropy', optimizer='rmsprop', metrics=['accuracy'])

# Upload images from your local machine to Colab
uploaded = files.upload()

# Define labels for Rock Paper Scissors based on alphabetical order, as used by ImageDataGenerator
labels = ['Paper', 'Rock', 'Scissors']

# Iterate over each uploaded file
for fn in uploaded.keys():
  # Construct the full path to the image file
  path = fn
  # Load the image and resize it to 150x150 pixels
  img = image.load_img(path, target_size=(150, 150))
  # Convert the image to a NumPy array
  x = image.img_to_array(img)
  # Rescale pixel values to be between 0 and 1 (important for models trained with ImageDataGenerator)
  x = x / 255.0
  # Expand dimensions to create a batch of 1 image (required by Keras models for prediction)
  x = np.expand_dims(x, axis=0)

  # Make a prediction using the trained model
  # Using 'x' directly as it is already a batch of 1 image
  classes = model.predict(x, batch_size=10)

  # Print the filename of the image
  print(fn)
  # Print the prediction probabilities for the image
  print(classes)

  # Get the predicted class index (the one with the highest probability)
  predicted_class_index = np.argmax(classes[0])
  # Print the predicted label using the defined 'labels' list
  print(f"{fn} is a {labels[predicted_class_index]}")


/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Saving image1 to image1
Saving image2 to image2
1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 773ms/step
image1
[[0.34299326 0.32582542 0.33118135]]
image1 is a Paper
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 36ms/step
image2
[[0.33402318 0.32922727 0.33674958]]
image2 is a Scissors


You can see from the filenames what the images were. Paper1.png ended up as [1, 0,
0], meaning the first neuron was activated and the others weren’t. Similarly,
Rock1.png ended up as [0, 1, 0], activating the second neuron, and Scissors2.png
was [0, 0, 1]. Remember that the neurons are in alphabetical order by label!
Some images that you can use to test the dataset are available to download. Alterna‐
tively, of course, you can try your own. Note that the training images are all done
against a plain white background, though, so there may be some confusion if there is
a lot of detail in the background of the photos you take.



Dropout Regularization
Earlier in this chapter we discussed overfitting, where a network may become too
specialized in a particular type of input data and fare poorly on others. One technique
to help overcome this is use of dropout regularization.
When a neural network is being trained, each individual neuron will have an effect
on neurons in subsequent layers. Over time, particularly in larger networks, some
neurons can become overspecialized—and that feeds downstream, potentially caus‐
ing the network as a whole to become overspecialized and leading to overfitting.
Additionally, neighboring neurons can end up with similar weights and biases, and if
not monitored this can lead the overall model to become overspecialized to the fea‐
tures activated by those neurons.
For example, consider the neural network in Figure 3-21, where there are layers of 2,
6, 6, and 2 neurons. The neurons in the middle layers might end up with very similar
weights and biases.
This reduces the chances of the neurons becoming overspecialized. The network will
still learn the same number of parameters, but it should be better at generalization—
that is, it should be more resilient to different inputs.
The concept of dropouts was proposed by Nitish Srivastava et al. in
their 2014 paper “Dropout: A Simple Way to Prevent Neural Net‐
works from Overfitting”.
To implement dropouts in TensorFlow, you can just use a simple Keras layer like this:


In [ ]:
import tensorflow as tf
tf.keras.layers.Dropout(0.2), # Dropout layer with a dropout rate of 0.2 (20% of neurons will be randomly set to 0 during training to prevent overfitting).

(<Dropout name=dropout, built=True>,)

For a simple example that demonstrates this, consider the Fashion MNIST classifier
from Chapter 2. I’ll change the network definition to have a lot more layers, like this:

In [ ]:
model = tf.keras.models.Sequential([
tf.keras.Input(shape=(28,28)),
tf.keras.layers.Flatten(),
tf.keras.layers.Dense(256, activation=tf.nn.relu),
tf.keras.layers.Dense(128, activation=tf.nn.relu),
tf.keras.layers.Dense(64, activation=tf.nn.relu),
tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])

Training this for 20 epochs gave around 94% accuracy on the training set, and about
88.5% on the validation set. This is a sign of potential overfitting.

In [ ]:
model = tf.keras.models.Sequential([
tf.keras.Input(shape=(28,28)),
tf.keras.layers.Flatten(),
tf.keras.layers.Dense(256, activation=tf.nn.relu),
tf.keras.layers.Dropout(0.2),
tf.keras.layers.Dense(128, activation=tf.nn.relu),
tf.keras.layers.Dropout(0.2),
tf.keras.layers.Dense(64, activation=tf.nn.relu),
tf.keras.layers.Dropout(0.2),
tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])

When this network was trained for the same period on the same data, the accuracy
on the training set dropped to about 89.5%. The accuracy on the validation set stayed
about the same, at 88.3%. These values are much closer to each other; the introduc‐
tion of dropouts thus not only demonstrated that overfitting was occurring, but also
that using dropouts can help remove such ambiguity by ensuring that the network
isn’t overspecializing to the training data.
Keep in mind as you design your neural networks that great results on your training
set are not always a good thing. This could be a sign of overfitting. Introducing
dropouts can help you remove that problem, so that you can optimize your network
in other areas without that false sense of security.

Summary
This chapter introduced you to a more advanced way of achieving computer vision
using convolutional neural networks. You saw how to use convolutions to apply fil‐
ters that can extract features from images, and designed your first neural networks to
deal with more complex vision scenarios than those you encountered with the
MNIST and Fashion MNIST datasets. You also explored techniques to improve your
network’s accuracy and avoid overfitting, such as the use of image augmentation and
dropouts.
Summary | 65
Before we explore further scenarios, in Chapter 4 you’ll get an introduction to
TensorFlow Datasets, a technology that makes it much easier for you to get access to
data for training and testing your networks. In this chapter you were downloading
ZIP files and extracting images, but that’s not always going to be possible. With
TensorFlow Datasets you’ll be able to access lots of datasets with a standard API

CHAPTER 4 : Using public datasets With Tensorflow Datasets


Image
From simple learning datasets like Horses or Humans up to advanced research
datasets for uses such as diabetic retinopathy detection
Object detection
COCO, Open Images, and more
Structured data
Titanic survivors, Amazon reviews, and more
Summarization
News from CNN and the Daily Mail, scientific papers, wikiHow, and more
Text
IMDb reviews, natural language questions, and more
Translate
Various translation training datasets
Video
Moving MNIST, Starcraft, and more

In [1]:
#installing tensorflow datasets
!pip install tensorflow-datasets